### PR TITLE
[ISSUE #1199] Implement the 1.0.0 openmessaging consumer API for rocketmq oms module

### PR DESCRIPTION
**Pull request by odbozhou**
## What is the purpose of the change

https://github.com/apache/rocketmq/issues/1199 Implement the 1.0.0 openmessaging consumer API for rocketmq oms module

## Brief changelog

1、Implement push consumer
2、Implement pull consumer

## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java [hunk 1]
```diff
@@ -76,110 +93,268 @@ public PullConsumerImpl(final KeyValue properties) {
...
+        properties.put("TIMEOUT", timeout);
+        MessageExt rmqMsg = localMessageCache.poll(properties);
         return rmqMsg == null ? null : OMSUtil.msgConvert(rmqMsg);
     }
 
     @Override
-    public Message receive(final KeyValue properties) {
+    public Message receive(String queueName, int partitionId, long receiptId, long timeout) {
+        KeyValue properties = new DefaultKeyValue();
+        properties.put("QUEUE_NAME", queueName);
```
Comment (zongtanghu) [inline openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java]: Here,the same as above.

File: openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java [hunk 2]
```diff
@@ -76,110 +93,268 @@ public PullConsumerImpl(final KeyValue properties) {
...
+        properties.put("RECEIPT_ID", receiptId);
+        properties.put("TIMEOUT", timeout);
         MessageExt rmqMsg = localMessageCache.poll(properties);
         return rmqMsg == null ? null : OMSUtil.msgConvert(rmqMsg);
     }
 
     @Override
-    public void ack(final String messageId) {
-        localMessageCache.ack(messageId);
+    public List<Message> batchReceive(long timeout) {
```
Comment (zongtanghu) [inline openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java]: In the batchReceive and localMessageCache.batchPoll  method,the timeout parameter won't have any effect.

File: openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java [hunk 3]
```diff
@@ -76,110 +93,268 @@ public PullConsumerImpl(final KeyValue properties) {
...
+                if (messageQueue.getQueueId() == partitionId) {
+                    mq = messageQueue;
+                }
+            }
+        } catch (MQClientException e) {
+            e.printStackTrace();
+        }
+        PullResult pullResult;
+        try {
+            pullResult = rocketmqPullConsumer.pull(mq, "*", receiptId, 4 * 1024 * 1024, timeout);
```
Comment (zongtanghu) [inline openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java]: Here,"4 * 1024 * 1024" you can add this contents into constants class.

File: openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java [hunk 4]
```diff
@@ -54,15 +57,76 @@ public Producer createProducer() {
     }
 
     @Override public Consumer createConsumer() {
-        return null;
+        String consumerId = accessPointProperties.getString("CONSUMER_ID");
```
Comment (zongtanghu) [inline openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java]: Here,do you want to distinguish pullConsumer from pushConsumer by the "CONSUMER_ID" key related value?And do you want set what's string value by using messagingAccessPoint.createPullConsumer method?

File: openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java [hunk 5]
```diff
@@ -54,15 +57,76 @@ public Producer createProducer() {
...
+        String[] nsStrArr = consumerId.split("_");
+        if (nsStrArr.length < 2) {
+            return new PushConsumerImpl(accessPointProperties);
+        }
+        if ("pull".equals(nsStrArr[0])) {
```
[the remark this answers is not in the document]
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java]: It has been modified

File: openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java [hunk 6]
```diff
@@ -54,15 +57,76 @@ public Producer createProducer() {
...
 
     @Override public MessageFactory messageFactory() {
         return null;
     }
+
+    class DefaultResourceManager implements ResourceManager {
+
+        @Override
+        public void createNamespace(String nsName) {
+            accessPointProperties.put("CONSUMER_ID", nsName);
```
[the remark this answers is not in the document]
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java]: It has been modified

File: openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java [hunk 7]
```diff
@@ -54,15 +57,76 @@ public Producer createProducer() {
...
+        }
+
+        @Override
+        public void deleteNamespace(String nsName) {
+            accessPointProperties.put("CONSUMER_ID", null);
```
[the remark this answers is not in the document]
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java]: It has been modified

File: openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java [hunk 8]
```diff
@@ -54,15 +57,76 @@ public Producer createProducer() {
...
+        }
+
+        @Override
+        public void switchNamespace(String targetNamespace) {
+            accessPointProperties.put("CONSUMER_ID", targetNamespace);
```
[the remark this answers is not in the document]
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java]: It has been modified

File: openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java [hunk 9]
```diff
@@ -54,15 +57,76 @@ public Producer createProducer() {
...
+        }
+
+        @Override
+        public Set<String> listNamespaces() {
+            return new HashSet<String>() {
+                {
+                    add(accessPointProperties.getString("CONSUMER_ID"));
```
[the remark this answers is not in the document]
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java]: It has been modified

File: openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/LocalMessageCache.java [hunk 10]
```diff
@@ -96,8 +103,8 @@ MessageExt poll() {
 
     MessageExt poll(final KeyValue properties) {
         int currentPollTimeout = clientConfig.getOperationTimeout();
-        if (properties.containsKey(Message.BuiltinKeys.TIMEOUT)) {
-            currentPollTimeout = properties.getInt(Message.BuiltinKeys.TIMEOUT);
+        if (properties.containsKey("TIMEOUT")) {
```
[the remark this answers is not in the document]
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/LocalMessageCache.java]: It has been modified

File: openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java [hunk 11]
```diff
@@ -76,110 +93,268 @@ public PullConsumerImpl(final KeyValue properties) {
 
         String consumerId = OMSUtil.buildInstanceName();
         this.rocketmqPullConsumer.setInstanceName(consumerId);
-        properties.put(OMSBuiltinKeys.CONSUMER_ID, consumerId);
+        properties.put("CONSUMER_ID", consumerId);
```
[the remark this answers is not in the document]
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java]: It has been modified

File: openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java (hunk 1 above)
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java]: It has been modified

File: openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java (hunk 2 above)
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java]: It has been modified

File: openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java (hunk 3 above)
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/consumer/PullConsumerImpl.java]: It has been modified

File: openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java (hunk 4 above)
Comment (odbozhou) [reply openmessaging/src/main/java/io/openmessaging/rocketmq/MessagingAccessPointImpl.java]: This is a temporary treatment plan，when you discuss it, then deal with this problem. @zongtanghu @duhengforever

**Pull request merged**